### PR TITLE
Let _get_dtype accept Categoricals and CategoricalIndex 

### LINKED ISSUE #16817: BUG: _get_dtype accepts CategoricalDtype but throws TypeError with a Categorical
#### Problem description

I find it inconistent that ``_get_dtype`` accept a ``CategoricalDtype``, but will thow TypeError on ``Categorical``s. 

#### Code Sample, a copy-pastable example if possible

```python
>>> cats = pd.Categorical([1,2,3])
>>> pd.core.dtypes.common._get_dtype(cats.dtype)
category
>>> pd.core.dtypes.common._get_dtype(cats)
TypeError: data type not understood
```

#### Expected Output

```python
>>> cats = pd.Categorical([1,2,3])
>>> pd.core.dtypes.common._get_dtype(cats.dtype)
category
>>> pd.core.dtypes.common._get_dtype(cats)
category
```

#### Proposed solution

A solution could be to check for extension type (maybe just categorical dtype, not sure) + that ``arr_or_dtype`` has a dtype attribute:

```python
def _get_dtypes(arr_or_dtype):
   ... 
    elif isinstance(arr_or_dtype, IntervalDtype):
        return arr_or_dtype   # old code
   elif is_extension_type(arr_or_dtype) and hasattr(arr_or_dtype, 'dtype'):   # new code
        return arr_or_dtype.dtype   # new code
    elif isinstance(arr_or_dtype, string_types):   # old code
      ...
```

### PR DESCRIPTION
**Pull request by topper-123**
- [ x] closes #16817
 - [x ] tests added / passed
 - [x ] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff`` (On Windows, ``git diff upstream/master -u -- "*.py" | flake8 --diff`` might work as an alternative.)
 - [ ] whatsnew entry

Makes ``_get_dtype`` accept ``Categorical``s and ``CategoricalIndex``. Also passes relevant tests in ``test_common.py``.

See also discussion in #16845. Is there a manual for checking performance (can't imagine this affects performance, though)?

### CODE REVIEW

**Comment by jreback** [comment]
> See also discussion in #16845. Is there a manual for checking performance (can't imagine this affects performance, though)?

This is in a critical path for lots of things. This absolutely needs checking.

http://pandas.pydata.org/pandas-docs/stable/contributing.html#running-the-performance-test-suite

**Comment by topper-123** [comment]
Alright, I've run asv. It took 2 hours, and the results are all over the place, from ratios of over 3 down to 0.01.  I assume this is not normal.

I tried selectively running the worst offenders individually, and then the results were now within limits (``BENCHMARKS NOT SIGNIFICANTLY CHANGED``). Do I need to take special precautions when running the whole suite, to avoid affecting the results?

File: pandas/tests/dtypes/test_common.py [hunk 1]
```diff
@@ -532,16 +532,16 @@ def test_is_complex_dtype():
     (float, np.dtype(float)),
     ('float64', np.dtype('float64')),
     (np.dtype('float64'), np.dtype('float64')),
-    pytest.mark.xfail((str, np.dtype('<U')), ),
+    (str, np.dtype(str)),
```
Comment (jreback) [inline pandas/tests/dtypes/test_common.py]: what fixed this one?

**Review by jreback — CHANGES_REQUESTED** [review]
add a whatsnew note in bug fix conversion section

File: pandas/tests/dtypes/test_common.py (hunk 1 above)
Comment (topper-123) [inline pandas/tests/dtypes/test_common.py]: On both my system (Windows) and at try.jupyter.org (sys.platform is 'linux', ``_get_dtype(str)`` returns ``np.dtype('<U')``. In the jenkins/appvoyeur,  it returns ``np.dtype('S')``. 

To me it looks like a difference of how numpy is handled on different system and not a pandas issue. I therefore relax that test, to make it only check that ``_get_dtype(str) == np.dtype(str)``, which passes both locally and in the CI system.

Comment (jreback) [inline pandas/tests/dtypes/test_common.py]: hmm, that looks like a py2/py3 thing. ok then. (you can maybe add a comment for that particular case)

**Comment by jreback** [comment]
ok on the asv. thanks for running that. just needs a whatsnew note. ping on green.

**Review by jreback — APPROVED** [review]

File: doc/source/whatsnew/v0.21.0.txt [hunk 2]
```diff
@@ -147,7 +147,8 @@ Bug Fixes
 Conversion
 ^^^^^^^^^^
 
-
+- :func:`dtypes.common._get_dtype` now accepts ``Categorical``s and
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: no need for an external note here; this is an internal function

Comment (topper-123) [inline doc/source/whatsnew/v0.21.0.txt]: I don't quite understand. Above you say

> just needs a whatsnew note

Are you thinking of a different location for the note?

Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: not this doesn't need a note at all. This is not a user relevant change as ``_get_dtype`` is not user facing.

**Comment by jreback** [comment]
thanks @topper-123

**Pull request merged**